File: pset1/Credit/credit.c
#include <cs50.h>
#include <stdio.h>

int get_one_digit(long number, long position);
int get_two_digits(long number, long position);

// TODO:
// 1. Ask for credit card number (Check)
// 2. Starting with the second to last digit, get every other number in the credit card number (Check)
// 3. Multiply these numbers by 2 (Check)
// 4. If any of these numbers are double digits, separate each digit from eachother (Check)
// 5. Add everything calculated (Check)
// 6. Add the numbers not divided by 2 from step 3 to the sum from step 5 (Check)
// 7. Check if last number from the sum of step 6 is a 0 (Check)
// 8. Check card length and what the starting digits are (Check)
// 9. Print AMEX, MASTERCARD, VISA or INVALID depending on step 7 and step 8 (Check)
int main(void)
{
    long i;
    long j;
    int digit;
    int sum = 0;
    int even_position_sum = 0;
    int odd_position_sum = 0;

    long n = get_long("Credit Card Number: ");
    for (i = 10; i < n; i *= 100)
    {
        digit = get_one_digit(n, i);
        digit *= 2;
        if (digit >= 10)
        {
            digit = get_one_digit(digit, 10) + get_one_digit(digit, 1);
        }
        even_position_sum += digit;
    }

    for (j = 1; j < n; j *= 100)
    {
        digit = get_one_digit(n, j);
        odd_position_sum += digit;
    }
    sum = even_position_sum + odd_position_sum;

    sum = get_one_digit(sum, 1);

    if (sum != 0)
    {
        printf("INVALID\n");
    }
    else if (4 == get_one_digit(n, 1000000000000000) || 4 == get_one_digit(n, 1000000000000))
    {
        printf("VISA\n");
    }
    else if (51 == get_two_digits(n, 100000000000000)  || 52 == get_two_digits(n, 100000000000000)
             || 53 == get_two_digits(n, 100000000000000) || 54 == get_two_digits(n, 100000000000000) || 55 == get_two_digits(n, 100000000000000))
    {
        printf("MASTERCARD\n");
    }
    else if (37 == get_two_digits(n, 10000000000000) || 34 == get_two_digits(n, 10000000000000))
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

int get_one_digit(long number, long position)
{
    return (number % (position * 10)) / position;
}
int get_two_digits(long number, long position)
{
    return number / position;
}

File: pset6/sentimental-credit/credit.py
import sys


def main():
    c_num = input("Enter a credit card number: ").strip()
    if not is_valid(c_num):
        print('INVALID')
        sys.exit(1)
        
    if c_num[0] == '4' and (len(c_num) == 16 or len(c_num) == 13):
        print('VISA')
        
    elif (c_num[:2] == '51' or \
        c_num[:2] == '52' or \
        c_num[:2] == '53' or \
        c_num[:2] == '54' or \
        c_num[:2] == '55') and len(c_num) == 16:

        print('MASTERCARD')
        
    elif c_num[:2] == '34' or c_num[:2] == '37' and len(c_num) == 15:
        print('AMEX')
        
    else:
        print('INVALID')



def is_valid(c_num):
    c_sum = 0
    if len(c_num) % 2 == 1:

        try:
            for i, digit in enumerate(map(int, c_num)):
                if i % 2 == 1:
                    digit *= 2
                    if digit > 9:
                        digit = (digit % 10) + (digit // 10)


                c_sum += digit
        except ValueError:
            return False
    else:
        for i, digit in enumerate(map(int, c_num)):
            if i % 2 == 0:
                digit *= 2
                if digit > 9:
                    digit = (digit % 10) + (digit // 10)


            c_sum += digit


    return c_sum % 10 == 0


if __name__ == '__main__':
    main()